cpp/src/arrow/compute/kernels/scalar_nested.cc

```diff
@@ -87,6 +89,195 @@ Status GetListElementIndex(const ExecValue& value, T* out) {
+                          const offset_type next_offset, const ListSliceOptions* opts,
+                          const ArraySpan* list_values) {
+    auto value_builder = list_builder->value_builder();
+    auto cursor = offset;
+
+    RETURN_NOT_OK(list_builder->Append());
+    while (cursor < offset + (opts->stop.value() - opts->start)) {
+      if (cursor + opts->start >= next_offset) {
+        if constexpr (!std::is_same_v<BuilderType, FixedSizeListBuilder>) {
+          break;  // don't pad nulls for variable sized list output
+        }
+        RETURN_NOT_OK(value_builder->AppendNull());
+      } else {
+        RETURN_NOT_OK(
+            value_builder->AppendArraySlice(*list_values, cursor + opts->start, 1));
+      }
+      ++cursor;
+    }
+    return Status::OK();
+  }
+};
+
+Result<TypeHolder> MakeListSliceResolve(KernelContext* ctx,
+                                        const std::vector<TypeHolder>& types) {
+  const auto start = OptionsWrapper<ListSliceOptions>::Get(ctx).start;
+  const auto stop = OptionsWrapper<ListSliceOptions>::Get(ctx).stop;
+  const auto return_fixed_size_list =
+      OptionsWrapper<ListSliceOptions>::Get(ctx).return_fixed_size_list;
+  const auto list_type = checked_cast<const BaseListType*>(types[0].type);
+  if (return_fixed_size_list) {
+    if (!stop.has_value()) {
+      return Status::NotImplemented(
+          "Unable to produce FixedSizeListArray without `stop` being set.");
+    }
+    return TypeHolder(fixed_size_list(list_type->value_type(),
+                                      static_cast<int32_t>(stop.value() - start)));
+  } else {
+    // Returning large list if that's what we got in and didn't ask for fixed size
+    if (list_type->id() == Type::LARGE_LIST) {
+      return list_type;
+    }
+    return TypeHolder(list(list_type->value_type()));
+  }
+}
+
+template <typename InListType, template <typename...> class Functor>
+void AddListSliceKernels(ScalarFunction* func) {
+  for (const auto& index_type : IntTypes()) {
+    auto inputs = {InputType(InListType::type_id)};
+    auto output = OutputType{MakeListSliceResolve};
+    auto scalar_exec = GenerateInteger<Functor, InListType>({index_type->id()});
+    ScalarKernel kernel(inputs, output, std::move(scalar_exec),
+                        OptionsWrapper<ListSliceOptions>::Init);
+    kernel.null_handling = NullHandling::COMPUTED_NO_PREALLOCATE;
+    kernel.mem_allocation = MemAllocation::NO_PREALLOCATE;
+    DCHECK_OK(func->AddKernel(std::move(kernel)));
+  }
+}
+
+void AddListSliceKernels(ScalarFunction* func) {
+  AddListSliceKernels<ListType, ListSlice>(func);
+  AddListSliceKernels<LargeListType, ListSlice>(func);
+  AddListSliceKernels<FixedSizeListType, ListSlice>(func);
+}
+
+const FunctionDoc list_slice_doc(
+    "Slice list into a subset",
+    ("`lists` must have a list-like type.\n"
+     "Return subset array of the list, as either a variable\n"
```

Comment:
Maybe something like "For each list value of `lists`, compute a slice and return as a new list array (either a variable or fixed size list array depending on the options)"?